### PR TITLE
Feature/consistent control conventions

### PR DESCRIPTION
**Pull request by supertom01**
All the mathematical steps that convert a given velocity and angle into wheel speeds and vice versa have been updated in order to match the notation and logic that has been used on the [wiki](https://wiki.roboteamtwente.nl/technical/control/omnidirectional).

This code has been tested on the robot hardware and does not show any unexpected behavior.

### CODE REVIEW

File: Core/Inc/Util/control_util.h [hunk 1]
```diff
@@ -79,18 +79,27 @@ float WHEEL_REF_LIMIT; 							// [rad/s] Limit the maximum wheel reference to le
...
+	yaw = 3,		// The angle
+}robot_axes;
 
 typedef enum {
 	wheels_RF,	// The right front wheel
-	wheels_RB,	// The right back wheel
-	wheels_LB,	// The left back wheel
 	wheels_LF,	// The left front wheel
+	wheels_LB,	// The left back wheel
+	wheels_RB,	// The right back wheel
```
Comment (skaeringur97) [inline Core/Inc/Util/control_util.h:102]: I agree with this change, but be careful that if python utils, AI code etc sends or reads data to/from the wheels that you change it there also

**Review by skaeringur97 — APPROVED** [review]
Looks good! Just make sure it´s easy to find the reasoning for this on the wiki. Since this has been tested, I think it can be merged (as long as my other comment is checked first)

File: Core/Inc/Util/control_util.h (hunk 1 above)
Comment (supertom01) [reply Core/Inc/Util/control_util.h:102]: Ah yes indeed. I did not think about that.

Comment (supertom01) [reply Core/Inc/Util/control_util.h:102]: The compatibility with the new wheel order is updated in the basestation with the next PR: RoboTeamTwente/Basestation#29. 

For AI I could not find any references to direct wheel speeds

Comment (skaeringur97) [reply Core/Inc/Util/control_util.h:102]: If you want to be extra careful you can disconnect all wheels except one and read plot the wheel speeds for each wheel respectively.

File: Core/Inc/Control/stateControl.h [hunk 2]
```diff
@@ -73,24 +73,24 @@ float* stateControl_GetWheelRef();
 /**
  * Set the current state as the estimated state as calculated by stateEstimation
  * 
- * @param input The x, y, w and yaw speeds from stateEstimation
+ * @param input The u, v, w and yaw speeds from stateEstimation
  */
 void stateControl_SetState(float input[4]);
```
Comment (emielsteerneman) [inline Core/Inc/Control/stateControl.h]: Request to change `input` to `stateLocal`

File: Core/Src/Control/stateControl.c [hunk 3]
```diff
@@ -1,54 +1,58 @@
...
+static float stateGlobalRef[4] = {0.0f}; 
 
-// The wheel velocities to be achieved.
-static float wheelRef[4] = {0.0f, 0.0f, 0.0f, 0.0f};
+// The wheel velocities to be achieved [rad/s]
+static float wheelRef[4] = {0.0f};
 
-// The current x, y, w and yaw velocities.
-static float state[4] = {0.0f};
+// The current global u, v, w and yaw velocities.
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c:19]: This should say `local` I assume

File: Core/Src/Control/stateControl.c [hunk 4]
```diff
@@ -1,54 +1,58 @@
...
+#include "logging.h"
 
 ///////////////////////////////////////////////////// VARIABLES
 
 // The current status of the system.
 static PID_states status = off;
 
-// The PID values for x, y, w and yaw.
+// The PID values for u, v, w and yaw.
 static PIDvariables stateK[4];
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: Request to change this to `stateLocalK`

File: Core/Src/Control/stateControl.c [hunk 5]
```diff
@@ -1,54 +1,58 @@
...
- * Translates the velocity from a body perspective to wheel speeds.
+ * Translates the velocity from a local perspective to wheel speeds.
  * 
- * @param wheelSpeed The speed to be achieved for each wheel.
- * @param vel 		 The velocities to be achieved seen from the body perspective.
+ * @param wheelSpeed The speed to be achieved for each wheel [rad/s]
+ * @param localState The velocities to be achieved seen from the body perspective {vel_u, vel_v, vel_w} [m/s]
  */
-static void body2Wheels(float wheelSpeed[4], float state[3]);
+static void body2Wheels(float wheelSpeed[4], float localState[3]);
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: Please stick to `stateLocal / stateGlobal` or `localState / globalState`. My preference goes to the first.

File: Core/Src/Control/stateControl.c [hunk 6]
```diff
@@ -1,54 +1,58 @@
...
 /**
- * Translates the global coordinate frame to the local coordinate frame
+ * Translates the global coordinate frame to the local coordinate frame.
  * 
- * @param global 	The global coordinates
- * @param local 	The local coordinates
- * @param yaw 		Yaw
+ * @param global 	The global coordinates {vel_x, vel_y, vel_w, yaw}
+ * @param local 	The local coordinates {vel_u, vel_v, vel_w, yaw}
+ * @param angle 	Yaw
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c:44]: We should choose some good definitions for yaw and angle. We have two angles on the robot; The direction the robot is looking at, and the direction the robot is moving towards. This function description says that `angle` == `yaw` which is confusing, at least to me. Any idea if we already have clear definitions?

File: Core/Src/Control/stateControl.c [hunk 7]
```diff
@@ -1,54 +1,58 @@
...
  * Determines the desired wheel speeds given the desired velocities
  * 
- * @param state 			The current x, y, w and yaw speeds seen from the body
- * @param velRef 			The instructed global x, y, w and yaw speeds
+ * @param localState 		The current u, v, w and yaw speeds seen from the body
+ * @param stateGlobalRef 			The instructed global x, y, w and yaw speeds
  * @param velocityWheelRef 	The resulting wheel speeds that should be achieved for each wheel
  */
-static void velocityControl(float state[3], float velRef[4], float velocityWheelRef[4]);
+static void velocityControl(float localState[3], float stateGlobalRef[4], float velocityWheelRef[4]);
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: Same request, either `state...` or `...State`, with preference towards `state...`

File: Core/Src/Control/stateControl.c [hunk 8]
```diff
@@ -81,53 +100,53 @@ int stateControl_DeInit(){
...
+	stateGlobalRef[vel_y] = input[vel_y];
+	stateGlobalRef[vel_w] = input[vel_w];
+	stateGlobalRef[yaw] = input[yaw];
 }
 
 float* stateControl_GetWheelRef() {
 	return wheelRef;
 }
 
 void stateControl_SetState(float input[4]){
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: request to change to `stateLocal` or `_stateLocal` or smth

File: Core/Src/Control/stateControl.c [hunk 9]
```diff
@@ -81,53 +100,53 @@ int stateControl_DeInit(){
...
+	stateLocal[vel_v] = input[vel_v];
+	stateLocal[vel_w] = input[vel_w];
+	stateLocal[yaw] = input[yaw];
 }
 
 void stateControl_GetPIDGains(PIDvariables gains[4]){
-	gains[body_x].kP = stateK[body_x].kP;
-	gains[body_x].kI = stateK[body_x].kI;
-	gains[body_x].kD = stateK[body_x].kD;
+	gains[vel_u].kP = stateK[vel_u].kP;
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: I think I already mentioned this but if possible change to `stateLocal`

File: Core/Src/Control/stateControl.c [hunk 10]
```diff
@@ -141,75 +160,83 @@ void stateControl_useAbsoluteAngle(bool angularControl){
...
 ///////////////////////////////////////////////////// PRIVATE FUNCTION IMPLEMENTATIONS
 
 static void body2Wheels(float wheelSpeed[4], float vel[3]){
-	wheelSpeed[wheels_RF] = (vel[body_x] * cosFront + vel[body_y] * sinFront) / rad_wheel;
-	wheelSpeed[wheels_RB] = (vel[body_x] * cosBack + vel[body_y] * -sinBack) / rad_wheel;
-	wheelSpeed[wheels_LB] = (vel[body_x] * -cosBack + vel[body_y] * -sinBack) / rad_wheel;
-	wheelSpeed[wheels_LF] = (vel[body_x] * -cosFront + vel[body_y] * sinFront) / rad_wheel;
 
+	// Translate the local u, v, and omega velocities into wheel velocities.
+	wheelSpeed[wheels_RF] = vel[vel_u] * D[0] + vel[vel_v] * D[1];
```
Comment (emielsteerneman) [inline Core/Src/Control/stateControl.c]: Is it possible to replace these indices (0, 1, 2, ...) with Enums or defines?

File: Core/Src/robot.c [hunk 11]
```diff
@@ -496,11 +496,11 @@ void loop(void){
 		robotFeedback.ballSensorSeesBall = ballPosition.canKickBall;
 		robotFeedback.ballPos = ballSensor_isInitialized() ? (-.5 + ballPosition.x / 700.) : 0;
 
-		float vx = stateEstimation_GetState()[body_x];
-		float vy = stateEstimation_GetState()[body_y];
-		robotFeedback.rho = sqrt(vx*vx + vy*vy);
-		robotFeedback.angle = stateEstimation_GetState()[body_yaw];
-		robotFeedback.theta = atan2(vy, vx);
+		float vu = stateEstimation_GetState()[vel_u];
```
Comment (emielsteerneman) [inline Core/Src/robot.c]: I am not a fan of multiple function calls, nor that `stateLocal` leaves the `stateEstimation` file scope by returning a pointer. But, it's not critical.

**Review by emielsteerneman — CHANGES_REQUESTED** [review]
All in all, looks good! Using `local` and `global` as well as `u` and `v` makes the code much more readable. Nice that you also tested it. I have some minor requests if you have the time, plus one or two actual things that need fixing.

**Comment by supertom01** [comment]
It seems like that robot feedback is being incorrectly calculated at the moment. Hence, do not merge yet!

File: Core/Src/Control/stateControl.c (hunk 6 above)
Comment (supertom01) [reply Core/Src/Control/stateControl.c:44]: Yeah there is at this moment not a strict definition yet indeed.

So in the code the `yaw` should technically always indicate the local angle of the robot. 
Whereas the `angle` should indicate the global angle of the robot.

File: Core/Src/Control/stateControl.c (hunk 10 above)
Comment (supertom01) [reply Core/Src/Control/stateControl.c]: Not really, or at least I cannot find of any variable names that would any value to these indices.

File: Core/Src/Control/stateControl.c (hunk 6 above)
Comment (emielsteerneman) [reply Core/Src/Control/stateControl.c:44]: with angle, do you mean the facing direction or the driving direction?

Comment (supertom01) [reply Core/Src/Control/stateControl.c:44]: The angle as in the facing direction. 

If we are talking about any angle at all in robot code, we're actually always talking about our facing direction. The only location where the driving direction is used is the x- and y- decomposition section, happening straight away in `robot.c`. 

https://github.com/RoboTeamTwente/roboteam_microcontroller5.0/blob/dba96402ebf50d9e8b2bfea6c3d334adc2a7d709/Core/Src/robot.c#L174-L180

**Review by casdoorn — APPROVED** [review]
LGTM, in the future it might be nice to convert some of the float arrays into structs to remove magic indices and make sure that at compile time all function calls can be checked for correct inputs and prevent out-of-bounds indexing.
For D and Dinv you can use float[m][n] definitions to make them actual matrices, if that makes it easier to name things. Or use the multiplyMatrix function in body2wheels/wheels2body.

None of these are blockers and are more suggestions for further cleanup, so feel free to merge.

**Review by emielsteerneman — APPROVED** [review]
Lgtm

**Pull request merged**